### PR TITLE
[PATCH v3] linux-gen: pktio: support transmitting packet vectors using odp_queue_enq()

### PR DESCRIPTION
**Pull request by matiaselo**
V2:
- Increase out_discards statistics counter when packets are dropped (Jere)

### CODE REVIEW

File: platform/linux-generic/odp_packet_io.c [hunk 1]
```diff
@@ -932,34 +932,86 @@ static inline int pktin_recv_buf(pktio_entry_t *entry, int pktin_index,
...
+	num = odp_packet_vector_tbl(pktv, &pkt_tbl);
+	ODP_ASSERT(num > 0);
+	sent = odp_pktout_send(queue_fn->get_pktout(queue), pkt_tbl, num);
+
+	/* Return success if any packets were sent. Free the possible remaining
+	   packets in the vector. */
+	if (odp_unlikely(sent <= 0))
+		return -1;
+	else if (odp_unlikely(sent != num))
+		odp_packet_free_multi(&pkt_tbl[sent], num - sent);
```
Comment (JereLeppanen) [inline platform/linux-generic/odp_packet_io.c:950]: In this case, aren't we dropping packets without any errors or counters visible to the application?

Comment (matiaselo) [reply platform/linux-generic/odp_packet_io.c:950]: When using odp_queue_enq() to send packets there may be intermediate steps (e.g. internal reorder queue) before the packets are actually transmitted to the wire, so trusting to the function return value is alone is difficult. The correct error counters should be incremented on the pktio driver level. I'll go through the drivers and add counters where necessary.

**Comment by matiaselo** [comment]
@JereLeppanen is V2 looking acceptable to you?

**Comment by JereLeppanen** [comment]
@matiaselo Yes, this looks good, approving.

**Review by JereLeppanen — APPROVED** [review]

**Pull request merged**